coordinator/coordinator.go

```diff
@@ -35,34 +36,81 @@ func (c *Coordinator) Run() error {
 		c.cluster.Shutdown()
 		return err
 	}
+	defer c.cluster.Shutdown()
```

Comment:
Why don't we move this above the Run call ? Then we can even remove the Shutdown call from inside the if condition.